File: 2015/Day20/main.py
def find_lowest_house(target_presents):
    house_presents = [0] * (target_presents // 10 + 1)
    for elf in range(1, len(house_presents)):
        for house in range(elf, len(house_presents), elf):
            house_presents[house] += elf * 10
    for house_number in range(1, len(house_presents)):
        if house_presents[house_number] >= target_presents:
            return house_number

def find_lowest_house2(target_presents):
    max_houses = target_presents // 11 + 1
    house_presents = [0] * max_houses
    for elf in range(1, len(house_presents)):
        for house in range(elf, min(elf * 50, len(house_presents)), elf):
            house_presents[house] += elf * 11
    for house_number in range(1, len(house_presents)):
        if house_presents[house_number] >= target_presents:
            return house_number

input = 34000000

print(f"Part 1: {find_lowest_house(input)}")
print(f"Part 2: {find_lowest_house2(input)}")

File: 2015/Day20/main.lua
local function find_lowest_house(target_presents)
    local max_houses = math.ceil(target_presents / 10)
    local house_presents = {}
    for i = 1, max_houses + 1 do
        house_presents[i] = 0
    end

    for elf = 1, #house_presents do
        for house = elf, #house_presents, elf do
            house_presents[house] = house_presents[house] + elf * 10
        end
    end

    for house_number = 1, #house_presents do
        if house_presents[house_number] >= target_presents then
            return house_number
        end
    end

    return -1 -- No house found
end

local function find_lowest_house2(target_presents)
    local max_houses = math.ceil(target_presents / 11)
    local house_presents = {}
    for i = 1, max_houses + 1 do
        house_presents[i] = 0
    end

    for elf = 1, #house_presents do
        for house = elf, math.min(50 * elf, #house_presents), elf do
            house_presents[house] = house_presents[house] + elf * 11
        end
    end

    for house_number = 1, #house_presents do
        if house_presents[house_number] >= target_presents then
            return house_number
        end
    end

    return -1 -- No house found
end

local input = 34000000

print('Part 1: ' .. find_lowest_house(input))
print('Part 2: ' .. find_lowest_house2(input))
